nemo/collections/nlp/models/nlp_model.py

```diff
@@ -312,53 +252,225 @@ def setup(self, stage: str) -> None:
-                        rank=app_state.data_parallel_rank,
-                    )
-                    mp_dl = self._trainer.replace_sampler(self._train_dl, mp_sampler)
-                    self._train_dl = mp_dl
-                else:
-                    raise NotImplementedError(
-                        f'The BERT encoder: {self.bert_model} does not support model parallelism yet.'
-                    )
-
     def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
         if hasattr(self, "bert_model") and isinstance(self.bert_model, MegatronBertEncoder):
             checkpoint['checkpoint_version'] = get_checkpoint_version()
         return None
 
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        if hasattr(self, "bert_model") and isinstance(self.bert_model, MegatronBertEncoder):
+            set_checkpoint_version(checkpoint['checkpoint_version'])
+            logging.info(f"Setting Megatron checkpoint version: {checkpoint['checkpoint_version']}")
+        return None
+
+    # no rank check as model parallel models need to be saved on data parallel rank 0
+    def save_to(self, save_path: str):
+        """
+        Saves model instance (weights and configuration) into .nemo file
+         You can use "restore_from" method to fully restore instance from .nemo file.
+
+        .nemo file is an archive (tar.gz) with the following:
+            model_config.yaml - model configuration in .yaml format. You can deserialize this into cfg argument for model's constructor
+            model_wights.chpt - model checkpoint
+
+        Args:
+            save_path: Path to .nemo file where model instance should be saved
+        """
+
+        app_state = AppState()
+        if app_state.model_parallel_size is not None:
+            self._default_save_to(save_path)
+        else:
+            return super().save_to(save_path)
+
+    def _default_save_to(self, save_path: str):
+        app_state = AppState()
+        if app_state.model_parallel_size is not None:
+            # each model parallel rank creates a .nemo file
+            # after all .nemo files are created, each rank
+            # will add their checkpoint to global rank 0
+
+            base_dir = os.path.dirname(save_path)  # use the directory to merge mp_rank .nemo files into one
+
+            # TODO: accept general paths
+            # update save_path based on model parallel_rank
+            base_path = save_path[0:-5]  # everything excpe the .nemo extension
+
+            mp_save_path = f'{base_path}_mp_rank_{app_state.model_parallel_rank:02d}.nemo'
+
+            if app_state.data_parallel_rank == 0:
+                super()._default_save_to(mp_save_path)
+
+            # barrier so that all processes have finished writing their weights before creating .nemo file
+            torch.distributed.barrier()
+
+            if is_global_rank_zero():
+                # TODO: do in a temporary directory
```

Comment:
I think (given the huge size of the models + the time it would take to 1) write to temp dir 2) copy from temp dir to required dir), this can be done in place. Upto you. Its cleaner to do in a temp dir, but with model parallel, we are talking about non trivial model checkpoint write times.

Reply:
That is a good point, with 4-8B parameter models that will be a huge copy. Let's leave it as is for now and we can revisit it later if needed.